### PR TITLE
[V1][Core] Fix memory issue with logits & sampling

### PR DESCRIPTION
**Pull request by ywang96**
Reopened from reverted #13776

Co-authored by @varun-sundar-rabindranath for LoRA dummy run fix.

### CODE REVIEW

**Comment by ywang96** [comment]
@varun-sundar-rabindranath @jeejeelee Please help take a look why this is breaking LoRA tests on V1 - thank you very much! 🙏

File: vllm/v1/worker/gpu_worker.py [hunk 1]
```diff
@@ -211,6 +213,25 @@ def compile_or_warm_up_model(self) -> None:
...
+
+        # Warm up sampler and preallocate memory buffer for logits and other
+        # sampling related tensors of max possible shape to avoid memory
+        # fragmentation issue.
+        # NOTE: This is called after `capture_model` on purpose to prevent
+        # memory buffers from being cleared by `torch.cuda.empty_cache`.
+        try:
+            self.model_runner._dummy_sampler_run(
+                hidden_states=self.model_runner._dummy_run(
+                    num_tokens=self.scheduler_config.max_num_seqs))
```
Comment (varun-sundar-rabindranath) [inline vllm/v1/worker/gpu_worker.py:225]: it looks like this should be `num_tokens=self.scheduler_config.max_num_batched_tokens` ?

Comment (varun-sundar-rabindranath) [reply vllm/v1/worker/gpu_worker.py:225]: looks like that alone is not enough.

Comment (ywang96) [reply vllm/v1/worker/gpu_worker.py:225]: My throught here is that during inference time, we only run sampling on the generated logits, so in the worst case scenario we will have `max_num_seqs` (all sequences in the batch are decoding) tokens.

**Comment by varun-sundar-rabindranath** [comment]
Have a PR #14514 that should fix this issue 🤞 @ywang96 please Cherry pick the latest commit to this branch. Thanks 🙏🏻

**Comment by varun-sundar-rabindranath** [comment]
Have a PR #14514 that should fix this issue 🤞 @ywang96 please Cherry pick the latest commit to this branch. Thanks 🙏🏻

**Comment by ywang96** [comment]
> Have a PR #14514 that should fix this issue 🤞 @ywang96 please Cherry pick the latest commit to this branch. Thanks 🙏🏻

Tested with some of your changes locally and now the lora tested indeed passes. Thanks for the help!

File: tests/basic_correctness/test_cumem.py [hunk 2]
```diff
@@ -142,7 +142,16 @@ def test_end_to_end(model: str, use_v1: bool):
     used_bytes = total - free_gpu_bytes_after_sleep - used_bytes_baseline
     # now the memory usage is mostly cudagraph memory pool,
     # and it should be less than the model weights (1B model, 2GiB weights)
-    assert used_bytes < 2 * GiB_bytes
+
+    # NOTE: In V1, the memory buffer for logits (max_num_reqs x vocab_size)
+    # is captured but cannot be releasesd from PyTorch due to a known bug,
```
Comment (WoosukKwon) [inline tests/basic_correctness/test_cumem.py:147]: Could you please elaborate on this?

Comment (ywang96) [reply tests/basic_correctness/test_cumem.py:147]: See the discussion here https://vllm-dev.slack.com/archives/C087WBWC5AQ/p1741398800083509?thread_ts=1741386694.452939&cid=C087WBWC5AQ

Comment (WoosukKwon) [reply tests/basic_correctness/test_cumem.py:147]: Hmm... Why do we need `empty_cache`?

Comment (ywang96) [reply tests/basic_correctness/test_cumem.py:147]: The difference here is that we never (in both V0 and V1) warmed up sampler, therefore the memory fragmentation issue was always there but not as pronounced in V0 (since the default batch size is 256).

Now we're adding the sampler warmup in V1, but when we call `sleep()`, the memory buffer for logits can't be cleared from the pytorch caching allocator (the bug mentioned in this comment), therefore the memory usage will be a lot higher.

Comment (WoosukKwon) [reply tests/basic_correctness/test_cumem.py:147]: @ywang96 Thanks for the explanation. Just want to double check: We don't want to call `empty_cache` anyways, because we intentionally reserve the `(max_num_reqs x vocab_size)`-sized tensor in the pytorch allocator, right?

Comment (ywang96) [reply tests/basic_correctness/test_cumem.py:147]: That is correct though I don think there should be a better & clean fix for this to work with sleep mode in the long term. We should probably free the memory when `sleep` is called, then warm up sampler again within `wakeup`, but this is currently blocked since we can free the memory anyways.

Comment (WoosukKwon) [reply tests/basic_correctness/test_cumem.py:147]: Hmm,,, How is the logits tensor different from other intermediate activation tensors?

Comment (ywang96) [reply tests/basic_correctness/test_cumem.py:147]: Because `dummy_run` doesn't include/activate sampler tensors, this is why we made `dummy_sampler_run` in the first place.

**Review by WoosukKwon — APPROVED** [review]
